components/zip321-parse/src/lib.rs

```diff
@@ -0,0 +1,924 @@
+/// transparent payment address.
+pub fn to_payment(vs: Vec<Param>, i: usize) -> Result<Payment> {
+    let recipient_address = vs
+        .iter()
+        .find_map(|v| match v {
+            Param::Addr(a) => Some(a.clone()),
+            _otherwise => None,
+        })
+        .context(RecipientMissingSnafu { idx: i })?;
+
+    let mut payment = Payment {
+        recipient_address,
+        amount_coins: 0,
+        amount_zatoshis: 0,
+        memo: None,
+        label: None,
+        message: None,
+        other_params: vec![],
+    };
+
+    for v in vs {
+        match v {
+            Param::Amount { coins, zatoshis } => {
+                payment.amount_coins = coins;
+                payment.amount_zatoshis = zatoshis;
+            }
+            Param::Memo(m) => {
+                payment.memo = Some(m);
+            }
+            Param::Label(m) => payment.label = Some(m),
+            Param::Message(m) => payment.message = Some(m),
+            Param::Other(n, m) => payment.other_params.push((n, m)),
+            _otherwise => {}
+        }
+    }
+
+    Ok(payment)
+}
+
+/// A single payment being requested.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Payment {
+    // TODO(schell): changed
+    /// The address to which the payment should be sent.
+    recipient_address: String,
+    // TODO(schell): changed, make separate types for whole and fractional
+    /// The amount of the payment that is being requested.
+    amount_coins: u64,
+    amount_zatoshis: u64,
+    // TODO(schell): changed
+    /// A memo that, if included, must be provided with the payment.
+    /// If a memo is present and [`recipient_address`] is not a shielded
+    /// address, the wallet should report an error.
+    ///
+    /// [`recipient_address`]: #structfield.recipient_address
+    memo: Option<Memo>,
+    /// A human-readable label for this payment within the larger structure
+    /// of the transaction request.
+    label: Option<String>,
+    /// A human-readable message to be displayed to the user describing the
+    /// purpose of this payment.
+    message: Option<String>,
+    /// A list of other arbitrary key/value pairs associated with this payment.
+    other_params: Vec<(String, String)>,
+}
+
+impl AsRef<Payment> for Payment {
+    fn as_ref(&self) -> &Payment {
+        self
+    }
+}
+
+impl AsMut<Payment> for Payment {
+    fn as_mut(&mut self) -> &mut Payment {
+        self
+    }
+}
+
+impl Payment {
+    /// Constructs a new [`Payment`] from its constituent parts.
+    pub fn new(
+        recipient_address: String,
+        amount: (u64, u64),
+        memo: Option<Memo>,
+        label: Option<String>,
+        message: Option<String>,
+        other_params: Vec<(String, String)>,
+    ) -> Self {
+        Self {
+            recipient_address,
+            amount_coins: amount.0,
+            amount_zatoshis: amount.1,
+            memo,
+            label,
+            message,
+            other_params,
+        }
```

Comment:
Notice here that there's no check to ensure the recipient address can receive a memo. This is by design since we don't have enough information to do this. The check still happens in `zip321::Payment::new`.